### DISCUSSION TITLE
Is there a way to add / edit the password of an entry from a batch file (Windows)?

### DISCUSSION THREAD

**Discussion by harrandt** [keepassxreboot/keepassxc#9898] [Q&A]
I would like to create passwords for devices without user interaction and store them in keypass.

I tried something like

`(echo mainpassword|keepassxc-cli.exe edit "C:\test.kdbx" MYUSER -p) < file_with_password.txt`

but it did not work. 
Another idea was to use interactive mode

`keepassxc-cli.exe open "C:\test.kdbx" < file_with_commands.txt`

with the contents of the file being

```
mainpassword
ls

```
but also without success. The login with the password works fine, but the commands are dropped.

The "ls" was just to see if it works at all, but the interactive mode does not respond to any command.

Any ideas how this could be done?

**Comment by droidmonkey** [comment]
You need to include everything into the echo pipe

**Reply by harrandt** [reply]
I tried that, but how would the command look like? I tried things like

`echo "mainpassword\ninteractivecommand1\ninteractivecommand2"|...`

**Reply by droidmonkey** [reply]
You might also remove the echo command and put everything in the text file. 

Or see here: https://stackoverflow.com/a/132811

**Reply by harrandt** [reply]
As I wrote above, I already tried putting everything in a text file, without success. I also tried the echo variations from the link you posted, again. No success.

The command I use is

`"c:\Program Files\KeePassXC\keepassxc-cli.exe" open "C:\test.kdbx" < "C:\keypass.in"`

The file content is

```
mypassword
ls
exit
```
The login works fine and the prompt shows

```
Please enter password to unlock C:\test.kdbx:
test>_
```

with a blinking but non-responsive cursor.

So the password is read from the file, but the following commands are not.
I even tried different newline sequences for Unix and Windows for the file.

The only thing I can do at the prompt is Ctrl-C, which does not show the commands from the file.
But on Ctrl-C it will show the letters entered before manually, not the one in the file.

```
keepassxc-cli.exe" --version
2.7.1
```

**Reply by droidmonkey** [reply]
The open command does not support passing stdin to the internal "shell"

**Reply by harrandt** [reply]
Ok, then how would it work using `echo`?

I tried many variations, but without success

```
(echo newuserpw && echo mainpassword)|"c:\Program Files\KeePassXC\keepassxc-cli.exe" edit "C:\test.kdbx" USER -p

echo newuserpw & echo mainpassword|"c:\Program Files\KeePassXC\keepassxc-cli.exe" edit "C:\test.kdbx" USER -p

echo newuserpw|(echo mainpassword|"c:\Program Files\KeePassXC\keepassxc-cli.exe" edit "C:\test.kdbx" USER -p)
```

Also with a separating `echo:` but without success. But something is happening, the commads above always remove
a set passwort for the user. So at the commands seem to accept a newline at some point, but an empty one.

**Reply by droidmonkey** [reply]
I can reproduce this, debugging it didn't yield any obvious issues. I wonder if there is just a weird thing with Batch that injects newline into STDIN when a pipe is opened?

**Reply by michaelk83** [reply]
If it's the same `echo` behavior as linux, it appends a newline by default. You'd need to use `echo -n` to avoid the trailing newline.

**Reply by harrandt** [reply]
In Windows it might work using what is mentioned here, https://stackoverflow.com/a/7105690/7919597 I will try that later
